Adding Clause Interleaving to method definitions (#14019)
lampepfl/dotty: docs/_docs/reference/other-new-features/generalized-method-syntax.md

---
layout: doc-page
title: "Generalized Method Syntax"
movedTo: https://docs.scala-lang.org/scala3/reference/other-new-features/generalized-method-syntax.html
---

> The inclusion of using clauses is not the only way in which methods have been updated, type parameter clauses are now allowed in any number and at any position.

Reviewer: I guess there should be some "experimental" marker at the top of this page.

## Syntax Changes

### In Scala 2

The old syntax only allowed zero or one type parameter clause, followed by any number of term clauses, optionnally followed by an implicit clause:

```scala

> def foo[T, U](x: T)(y: U)(z: Int, s: String)(using Ord[Int])(a: Array[T])(implicit l: List[U])

Reviewer: The section title is "In Scala 2" but there's a `using` clause here. That's not consistent.

```

### In Scala 3

The new syntax allows any number of type clauses, as long as they are not adjacent:
(do note however that [implicit clause are discouraged, in favor of using clauses](https://docs.scala-lang.org/scala3/reference/contextual/relationship-implicits.html))

```scala
def foo[T, U](x: T)(y: U)[V](z: V, s: String)(using Ord[Int])[A](a: Array[A])(implicit List[U])
```

### Unchanged

Class definitions and type declarations are unaffected, there can only be up to one type clause, in leading posion.

## Motivation

The new syntax is a powerful but natural extension of the old one, it allows new design patterns while staying intuitive and legible.

### Dependent Type Clauses

> As type clauses can come after term clauses, it is now possible to have type parameters depend on term parameters:

Reviewer: ```suggestion
As type clauses can come after term clauses, it is now possible to have type parameters that depend on term parameters:
```

```scala
trait Key { type Value }
trait DB {

>   def get(k: Key): Option[k.Value] // dependent return type

Reviewer: Being pedantic:
```suggestion
  def get(k: Key): Option[k.Value] // dependent result type
```
Java has return types. Scala has result types.

  def getOrElse(k: Key)[V >: k.Value](default: V): V // dependent type parameter
}
```

Note that simply replacing `V` by `k.Value` would not be equivalent. For example, if `k.Value` is `Int`, the above allows: 

> `getOrElse(k)[Number](0.1)`, which returns a `Number`.

Reviewer: `Number` is not a supertype of `Int`. It's only a supertype of `Integer`. You need a different example here.

Author: I think I'll use `Option` then, thank you for reminding me

## Details

### Application

> Method application is unchanged, in cases where multiple type clauses are expected but not all are passed, the rightmost ones are inferred.

Reviewer: ```suggestion
Method application is unchanged.
When multiple type clauses are expected but not all are passed, the rightmost ones are inferred.
```

In particular, the following does not type check, even though the argument `Char` is only valid for `C`:
```scala
def triple[I <: Int](using Ordering[I])[C <: Char](a: I, b: C) = ???
triple[Char](0,'c') // error: Char does not conform to upperbound Int
```

### Extension Methods

> Naturally extension methods follow the same syntax, so the following is valid:

Reviewer: ```suggestion
Extension methods follow the same syntax, so the following is valid:
```

```scala
extension [T](l1: List[T])
  def zipWith[U](l2: List[U])[V](l3: List[V]): List[(T,U,V)]
```

### When to use

We recommand to always put a unique type clause at the beginning, unless it is not possible to do so.
For example, the extension method `zipWith` above should be written `zipWith[U, V](l2: List[U], l3: List[V]): List[(T,U,V)]` instead.

> The `getOrElse` method on the other hand can not be written with a leading type clause, it is therefore as recommended as is.

Reviewer: ```suggestion
The `getOrElse` method on the other hand can not be written with a leading type clause.
```

---
layout: doc-page
title: "Generalized Method Syntax"
movedTo: https://docs.scala-lang.org/scala3/reference/other-new-features/generalized-method-syntax.html
---

The inclusion of using clauses is not the only way in which methods have been updated, type parameter clauses are now allowed in any number and at any position.

## Syntax Changes

### In Scala 2

The old syntax only allowed zero or one type parameter clause, followed by any number of term clauses, optionnally followed by an implicit clause:

```scala
def foo[T, U](x: T)(y: U)(z: Int, s: String)(a: Array[T])(implicit List[U])
```

### In Scala 3

> The new syntax allows any number of type, term and using clause, in any order, optionnaly followed by an implicit clause:

Reviewer 2: The new syntax should be marked as experimental first. It should be available under a compiler option, e.g. `-Yinterleaved-clauses`.

Reviewer 3: isn't an experimental language import more idiomatic?

Reviewer 4: It shouldn't be needing a special flag, IMO. The feature should just be available if `experimental` is imported.

### Dependent Type Clauses

As type clauses can come after term clauses, it is now possible to have type parameters depend on term parameters:

```scala
trait Key { type Value }
trait DB {
  def get(k: Key): Option[k.Value] // dependent return type
  def getOrElse(k: Key)[V >: k.Value](default: V): V // dependent type parameter
}
```

Note that simply replacing `V` by `k.Value` would not be equivalent. For example, if `k.Value` is `Int`, the above allows: 
`getOrElse(k)[Number](0.1)`, which returns a `Number`.

### Partial Inference

It is now possible to only infer some of the type parameters, this reduces boilerplate at the use site:
```scala
trait StaticSizeList[S <: Int & Singleton, T]
def filled[S <: Int & Singleton][T](x: T): StaticSizeList[S,T] = ???
val helloes = filled[4]("Hello!") // S=4, and T is inferred
```

## Details

### Application

Method application is unchanged, in cases where multiple type clauses are expected but not all are passed, the rightmost ones are inferred.

In particular, the following does not type check, even though the argument `Char` is only valid for `C`:
```scala

> def triple[A <: Int][B <: String][C <: Char](a: A, b: B, c: C) = ???

Reviewer 5: This example should be changed now that we don't allow repeated type applications, right?
